includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php

```diff
@@ -1078,30 +1078,36 @@ public function add_contact( $contact, $list_id = false ) {
 		try {
 			$mc             = new Mailchimp( $this->api_key() );
 			$update_payload = [ 'email_address' => $email_address ];
-			$merge_fields   = [];
-			if ( isset( $contact['name'] ) ) {
-				$name_fragments = explode( ' ', $contact['name'], 2 );
-				$merge_fields   = [
-					'FNAME' => $name_fragments[0],
-				];
-				if ( isset( $name_fragments[1] ) ) {
-					$merge_fields['LNAME'] = $name_fragments[1];
-				}
-				$update_payload['merge_fields'] = $merge_fields;
-			}
 
-			// Get list merge fields (metadata) to create them if needed.
-			if ( ! empty( $merge_fields ) ) {
-				$list_merge_fields = array_reduce(
-					$mc->get( "lists/$list_id/merge-fields", [ 'count' => 100 ] )['merge_fields'],
-					function( $acc, $field ) {
-						$acc[ $field['name'] ] = $field['tag'];
-						return $acc;
-					},
-					[]
-				);
-			}
 			if ( isset( $contact['metadata'] ) && is_array( $contact['metadata'] ) && ! empty( $contact['metadata'] ) ) {
+				$update_payload['merge_fields'] = [];
+
+				$existing_merge_fields = $mc->get( "lists/$list_id/merge-fields", [ 'count' => 1000 ] )['merge_fields'];
+				usort(
+					$existing_merge_fields,
+					function( $a, $b ) {
+						return $a['merge_id'] - $b['merge_id'];
+					}
+				);
+
+				$list_merge_fields = [];
+
+				// Handle duplicate fields.
+				foreach ( $existing_merge_fields as $key => $field ) {
+					if ( ! isset( $list_merge_fields[ $field['name'] ] ) ) {
```

Comment:
I could not reproduce based on the instructions provided. Isn't this here a constant condition? `$list_merge_fields` is defined above as an empty array, so `! isset( $list_merge_fields[ $field['name'] ] )` will always be true.

Reply:
That's intended. Here's the breakdown of the logic:

1. We [fetch](https://github.com/Automattic/newspack-newsletters/pull/1370/files/f25ec1e0a1e60bafff5140e28ab55c2b046ce425#diff-961258e170b9bfe6e81621bab727a9a65947cd67cc016bf5ce1926dd8c78d77cR1085) all merge fields from the ESP.
2. We [sort](https://github.com/Automattic/newspack-newsletters/pull/1370/files/f25ec1e0a1e60bafff5140e28ab55c2b046ce425#diff-961258e170b9bfe6e81621bab727a9a65947cd67cc016bf5ce1926dd8c78d77cR1086) the fields in order of ascending `merge_id`. MC increments the ID so it's a decent approximation of when fields were created relative to each other.
3. We [add](https://github.com/Automattic/newspack-newsletters/pull/1370/files/f25ec1e0a1e60bafff5140e28ab55c2b046ce425#diff-961258e170b9bfe6e81621bab727a9a65947cd67cc016bf5ce1926dd8c78d77cR1097) the fields to `$list_merge_fields` (which is essentially an array of the fields we actually want to sync data to), keyed by field name.
4. Because of the `! isset( $list_merge_fields[ $field['name'] ] )`, if we encounter another field with the exact same name (the issue we're trying to correct for), we do not add this field to `$list_merge_fields`. Instead, since it's a duplicate field, we [delete](https://github.com/Automattic/newspack-newsletters/pull/1370/files/f25ec1e0a1e60bafff5140e28ab55c2b046ce425#diff-961258e170b9bfe6e81621bab727a9a65947cd67cc016bf5ce1926dd8c78d77cR1100) the field in the ESP.

Maybe the testing instructions aren't as detailed as they need to be. I've amended them with a bit more detail so hopefully it's easier to replicate the duplicate field handling.

Reply:
Thank you for the thorough response, this is much clearer now. My comment about the constant condition was misguided, I failed to notice that the array will be updated during the iteration 🙈